File: 84.柱状图中最大的矩形.py
#
# @lc app=leetcode.cn id=84 lang=python
#
# [84] 柱状图中最大的矩形
#
# 解法1(超时): 对于每根柱子，把他作为高，向左向右找第一个比他矮的就找到了宽，整体遍历一遍找最大的面积就解决了
# 
#   改进(T92% S58%)：每次都向左向右找比他小的，整体复杂度就是$O(N^2)$了 → **单调栈**：依次进，当进某一个的时候发展它比栈顶小，那就意味着当前栈顶的下面和栈外都比它小了，而且都是第一个比他小的东西，这时把当前栈顶弹出来可以计算以它为高的最大面积了（左边界就是弹完当前元素后的栈顶，右边界是当前栈外的元素），且做完一个之后栈顶可能还是比外面的大，还可以继续while不停做。本质上相当于当来了一个比较小的元素时，它很可能是很多位置第一个最小的右边界，有了它的位置就可以计算很多根大
#   注意因为要获取的是宽度，所以单调栈里放的是下标

# @lc code=start
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        if len(heights)==0: return 0
        elif len(heights)==1: return heights[0]

        heights = [0] + heights + [0]

        stack = [0]
        maxArea = 0
        for i in range(1, len(heights)):
            while heights[i] < heights[stack[-1]]:     # 当前元素比栈顶小 -> 栈顶找到了左右第一个比它小的
                thisIdx = stack.pop(-1)
                area = heights[thisIdx] * (i-stack[-1]-1)
                if area > maxArea:
                    maxArea = area
            stack.append(i)

        return maxArea

    def otherSolution(self, heights):
        # 解法1 超时
        heights = [0] + heights + [0]
        
        maxArea = 0
        for i in range(1, len(heights)-1):
            for j in range(i-1, -1, -1):
                if heights[j] < heights[i]: break
            for k in range(i+1, len(heights)):
                if heights[k] < heights[i]: break
            
            area = heights[i] * (k-j-1)
            if area > maxArea:
                maxArea = area
        return maxArea

Solution().largestRectangleArea([2,1,5,6,2,3])
# @lc code=end

File: javascript/84.柱状图中最大的矩形.js
/*
 * @lc app=leetcode.cn id=84 lang=javascript
 *
 * [84] 柱状图中最大的矩形
 * 
 * 很快就想到了单调栈，应该就不难了，为什么一直卡住呢，我们再来捋下思路：
 * 维护一个单调栈（存的是下标，因为要求width算面积，只存height宽度就丢了），如果遍历到某元素时发现它比栈顶要小，由于栈底到栈顶递增，所以此时栈顶发现了左边第一个比他小的以及右边第一个比他小的
 * 这样问题就会到了暴力的双重循环法
 * 对于某个元素，如果找到了左边第一个比他小的和右面第一个比他小的，则以当前元素为height的最大面积就可以求解了
 * 
 * 比较坑的是边界的判断，当栈为空，也就意味着左面都比当前元素大，那0就应该是左边界
 * 同样的，如果右面一直比自己大，那找到右边界也还是没有结果，此时添加一个0元素在原数组尾部就可以解决
 */

// @lc code=start
/**
 * @param {number[]} heights
 * @return {number}
 */
// 前后都添加一个0可以更好的处理，因为stack永远不会为空，0总归比所有人都小，左边界一直在那
// 同时需要注意需要while来处理，如果发现一个很小的右边界，它可能是很多人的右边界，在栈里不断的判断是否可以计算
var largestRectangleArea = function(heights) {
    if (heights.length == 1) { return heights[0]; }

    heights.unshift(0);
    heights.push(0);

    var res = 0;
    var stack = [0]; // 单调栈，栈从下到上递增
    for (var i=1; i<heights.length; i++) {
        while (heights[i] < heights[stack[stack.length-1]]) {
            var top = stack.pop();
            res = Math.max(res, heights[top]*(i-stack[stack.length-1]-1))
        }
        stack.push(i);
    }
    
    return res;
};

var largestRectangleArea2 = function(heights) {
    if (heights.length == 1) { return heights[0]; }
    heights.push(0);

    var res = 0;
    var stack = [0]; // 单调栈，栈从下到上递增
    for (var i=1; i<heights.length; i++) {
        while (stack.length && heights[i] < heights[stack[stack.length-1]]) {
            var top = stack.pop();
            if (stack.length!=0) {
                res = Math.max(res, heights[top]*(i-stack[stack.length-1]-1));
            } else {
                res = Math.max(res, heights[top]*(i));
            }
        }
        stack.push(i);
    }
    
    return res;
};

// 超时 95/98 每个位置向左向右找第一个比自己小的，求得最大面积
var largestRectangleArea1 = function(heights) {
    var res = 0;
    for (var i=0; i<heights.length; i++) {
        var left=i-1, right=i+1;
        while (left > -1 && heights[left]>=heights[i]) {
            left--;
        }
        while (right < heights.length && heights[right]>=heights[i]) {
            right++;
        }
        res = Math.max(res, heights[i]*(right-left-1));
    }
    return res;
};
// @lc code=end

